lib/pos_snapshot_entries.go

```diff
@@ -0,0 +1,764 @@
+		if err == badger.ErrKeyNotFound {
+			return nil, nil
+		}
+		return nil, errors.Wrapf(err, "DBGetSnapshotGlobalParamsEntryWithTxn: problem retrieving value")
+	}
+
+	// Decode from bytes.
+	globalParamsEntry := &GlobalParamsEntry{}
+	rr := bytes.NewReader(globalParamsEntryBytes)
+	if exist, err := DecodeFromBytes(globalParamsEntry, rr); !exist || err != nil {
+		return nil, errors.Wrapf(err, "DBGetSnapshotGlobalParamsEntryWithTxn: problem decoding GlobalParamsEntry: ")
+	}
+	return globalParamsEntry, nil
+}
+
+func DBPutSnapshotGlobalParamsEntryWithTxn(
+	txn *badger.Txn,
+	snap *Snapshot,
+	globalParamsEntry *GlobalParamsEntry,
+	snapshotAtEpochNumber uint64,
+	blockHeight uint64,
+) error {
+	if globalParamsEntry == nil {
+		// This should never happen but is a sanity check.
+		glog.Errorf("DBPutSnapshotGlobalParamsEntryWithTxn: called with nil GlobalParamsEntry, this should never happen")
+		return nil
+	}
+	key := DBKeyForSnapshotGlobalParamsEntry(snapshotAtEpochNumber)
+	return DBSetWithTxn(txn, snap, key, EncodeToBytes(blockHeight, globalParamsEntry))
+}
+
+//
+// SnapshotValidatorEntry
+//
+
+type SnapshotValidatorMapKey struct {
+	SnapshotAtEpochNumber uint64
+	ValidatorPKID         PKID
+}
+
+func (bav *UtxoView) SnapshotCurrentValidators(snapshotAtEpochNumber uint64, blockHeight uint64) error {
+	// First, snapshot any !isDeleted ValidatorEntries in the UtxoView.
+	var utxoViewValidatorPKIDs []*PKID
+	for _, validatorEntry := range bav.ValidatorPKIDToValidatorEntry {
+		if !validatorEntry.isDeleted {
+			// We only want to snapshot !isDeleted ValidatorEntries.
+			bav._setSnapshotValidatorEntry(validatorEntry, snapshotAtEpochNumber)
+
+			// Check if we should jail the validator.
+			shouldJailValidator, err := bav.ShouldJailValidator(validatorEntry, blockHeight)
+			if err != nil {
+				return errors.Wrapf(
+					err,
+					"SnapshotValidators: problem determining if should jail validator %v: ",
+					validatorEntry.ValidatorPKID,
+				)
+			}
+			// Jail them if so.
+			if shouldJailValidator {
+				if err = bav.JailValidator(validatorEntry); err != nil {
+					return errors.Wrapf(
+						err, "SnapshotValidators: problem jailing validator %v: ", validatorEntry.ValidatorPKID,
+					)
+				}
+			}
+		}
+		// We don't want to retrieve any ValidatorEntries from the db that are present in the UtxoView.
+		utxoViewValidatorPKIDs = append(utxoViewValidatorPKIDs, validatorEntry.ValidatorPKID)
+	}
+	// Second, snapshot the ValidatorEntries in the db (skipping any in the UtxoView).
+	dbValidatorEntries, err := DBEnumerateAllCurrentValidators(bav.Handle, utxoViewValidatorPKIDs)
+	if err != nil {
+		return errors.Wrapf(err, "SnapshotValidators: problem retrieving ValidatorEntries: ")
+	}
+	for _, validatorEntry := range dbValidatorEntries {
+		bav._setSnapshotValidatorEntry(validatorEntry, snapshotAtEpochNumber)
+
+		// Check if we should jail the validator.
+		shouldJailValidator, err := bav.ShouldJailValidator(validatorEntry, blockHeight)
```

Comment:
It feels like jailing validators in the snapshotting code is a bit overloaded. I was imagining something more like the following, which un-comingles the jailing from the snapshotting.

```
func RunOnEpochCompeted() {
    // Compute isLastBlockInEpoch. Return if it's not.
    // Jail validators that haven't been active
    // Pay all the top stakers (sofonias will handle this)
    // Snapshot validators
    // Snapshot leader schedule
    // Snapshot stakers (sofonias will augment it to do this)
    // Snapshot GlobalParams
    // Update CurrentEpochEntry
}
```

There is another issue which is whether this ORDER of "jailing" -> "paying previously snapshotted people" -> "snapshotting new people" is "correct." I played with it a bit, and I think jailing first makes sense. But we also have to think of it in the context of validators joining and leaving the "active set" in the event that we introduce a ConsensusMaxNumValidators. In particular, if we only allow votes by the top 100 validators by stake, you could imagine a situation where validator 101 doesn't vote because he's not in the top set, but then his stake increases and he IS in the top set BUT hasn't voted. I think he actually woudln't be jailed in this case because he'll vote during the first epoch when he joins the active set, which will show that he's been active. But I'm just writing it out to illustrate the potential edge case of someone joining the active set.

But sofonias will be the one dealing with this anyway.